### PR TITLE
Issue #3314741: EntityFieldManager::getFieldMap() doesn't show farmOS bundle fields

### PR DESCRIPTION
**Pull request by mstenta**
See https://www.drupal.org/project/farm/issues/3314741

### CODE REVIEW

**Comment by mstenta** [comment]
Note that this PR currently includes two commits that change the `equipment` and `quick` fields from bundle fields to base fields.

If we want to split those out to review separately that's fine with me.

File: modules/core/entity/farm_entity.module [hunk 1]
```diff
@@ -14,6 +14,121 @@ use Drupal\entity\EntityPermissionProvider;
...
+  if (!\Drupal::moduleHandler()->hasImplementations('farm_entity_bundle_field_info', $modules)) {
+    return;
+  }
+
+  // Iterate through entity types.
+  $entity_type_definitions = \Drupal::service('entity_type.manager')->getDefinitions();
+  foreach ($entity_type_definitions as $entity_type => $entity_type_definition) {
+
+    // Only proceed for entity types that use bundle plugins.
+    if (!in_array($entity_type, ['asset', 'log', 'plan', 'quantity'])) {
```
Comment (symbioquine) [inline modules/core/entity/farm_entity.module:65]: I'm seeing [this array](https://github.com/farmOS/farmOS/blob/c92d181c45b2158b360318edb277ff175fb78cd7/modules/core/entity/farm_entity.module#L73) in a lot of places - sometimes in different orders or with/without the 'data_stream' entry.

Would it make sense to refactor it into a constant - or maybe just additional attributes on the entity definitions themselves?

e.g. here;

```php
if (!$entity_type_definition->supportsBundlePlugins()) {
    continue;
}
```

File: modules/core/entity/farm_entity.module [hunk 2]
```diff
@@ -14,6 +14,121 @@ use Drupal\entity\EntityPermissionProvider;
...
+            if (empty($bundle_field_map[$field_name]['bundles'])) {
+              unset($bundle_field_map[$field_name]);
+            }
+          }
+        }
+      }
+    }
+
+    // Set the bundle field map key value collection.
+    if (!empty($bundle_field_map)) {
```
Comment (symbioquine) [inline modules/core/entity/farm_entity.module:123]: What if it became empty as part of the changes above? Shouldn't it still get set in that case?

File: modules/core/entity/farm_entity.post_update.php [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    // Set the bundle field map key value collection.
+    if (!empty($bundle_field_map)) {
+      \Drupal::service('keyvalue')->get('entity.definitions.bundle_field_map')->set($entity_type, $bundle_field_map);
+    }
+  }
+
+  // Delete the entity field map cache entry.
+  \Drupal::service('cache.discovery')->delete('entity_field_map');
```
Comment (symbioquine) [inline modules/core/entity/farm_entity.post_update.php:71]: Why implement this all again? Couldn't it just call `_farm_entity_rebuild_bundle_field_map('install', $modules)` with a list of all the modules that implement `hook_farm_entity_bundle_field_info`?

File: modules/core/entity/farm_entity.post_update.php [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+/**
+ * @file
+ * Post update functions for farmOS entity module.
+ */
+
+/**
+ * Add hook_farm_entity_bundle_field_info() fields to bundle field maps.
+ */
+function farm_entity_post_update_add_farm_entity_bundle_field_maps(&$sandbox = NULL) {
```
Comment (symbioquine) [inline modules/core/entity/farm_entity.post_update.php:11]: What's this `$sandbox` parameter? Do we need to use/honor it?

Comment (symbioquine) [inline modules/core/entity/farm_entity.post_update.php:11]: Probably just my Drupal naivete showing here, but does this implement a hook or is it called some other way?

File: modules/core/entity/tests/src/Functional/FarmEntityBundleFieldTest.php [hunk 5]
```diff
@@ -59,6 +59,49 @@ protected function setUp():void {
...
+
+    // Reload the entity field map.
+    $this->entityFieldManager->setFieldMap([]);
+    \Drupal::service('cache.discovery')->delete('entity_field_map');
+    $field_map = $this->entityFieldManager->getFieldMap();
+
+    // Confirm that the 'test_contrib_hook_bundle_field' no longer exists in the
+    // log field map.
+    $this->assertArrayNotHasKey('test_contrib_hook_bundle_field', $field_map['log']);
+  }
```
Comment (symbioquine) [inline modules/core/entity/tests/src/Functional/FarmEntityBundleFieldTest.php:103]: Should this also test the case where more than one module provides a given bundle field? i.e. that the bundle gets added, the field mapping doesn't get overwritten or prematurely deleted?

File: modules/core/quick/tests/src/Kernel/QuickFormTest.php [hunk 6]
```diff
@@ -94,13 +95,13 @@ public function testQuickFormSubmission() {
     $this->assertNotEmpty($storage['assets'][0]->id());
 
     // Confirm that the asset is linked to the quick form.
-    $this->assertEquals('test', $storage['assets'][0]->quick[0]);
+    $this->assertEquals('test', $storage['assets'][0]->get('quick')[0]->value);
```
Comment (symbioquine) [inline modules/core/quick/tests/src/Kernel/QuickFormTest.php:98]: Why does this need to change?

File: modules/core/entity/farm_entity.module (hunk 1 above)
Comment (mstenta) [reply modules/core/entity/farm_entity.module:65]: Oh good idea! Didn't know about the `supportsBundlePlugins()` method!

**Review by paul121 — COMMENTED** [review]
Just to capture what we chatted about on the monthly call..

This reminded me of this issue which provides a solution for a problem I have run into in update hooks adding/removing a bundle field in a custom module: https://www.drupal.org/project/drupal/issues/3129179

So considering that, it would be nice if there was a helper function we could call at any time to rebuild this field map during update hooks. To simplify things this helper function could rebuild the entire mapping, regardless of modules installed/uninstalled.

This issue I referenced may provide a solution for Drupal core, but we may need to determine if it works for the farmOS bundle field hook. It sounds like providing our own helper method for farmOS core would be the easiest near term solution (and call this helper function to do this logic as is being done by this PR)

File: modules/core/entity/farm_entity.module (hunk 1 above)
Comment (symbioquine) [reply modules/core/entity/farm_entity.module:65]: image

Comment (paul121) [reply modules/core/entity/farm_entity.module:65]: I think we could use `EntityTypeInterface::getBundleEntityType()`

```

  /**
   * Gets the name of the entity type which provides bundles.
   *
   * @return string|null
   *   The name of the entity type which provides bundles, or NULL if the entity
   *   type does not have a bundle entity type.
   */
  public function getBundleEntityType();
 ```

**Comment by mstenta** [comment]
@paul121 raised an important question on the dev call today: if we convert the `equipment` field to a base field, will that stop it from showing up in Log Views?

**Comment by mstenta** [comment]
Ran into this again with the new Convention taxonomy + reference field provided by this module: https://github.com/mstenta/farm_convention/

Worked around it by converting it from a bundle field to a base field: https://github.com/mstenta/farm_convention/commit/a7fa1bcf8902a909e9f15276eaefea7a5cbe20de

**Comment by mstenta** [comment]
I'm going to close this PR and start a new one that is based off of farmOS 3.x. I will explore whether we can patch Drupal core based on the work being done in https://www.drupal.org/project/drupal/issues/3129179.

Thanks for the reviews and feedback @symbioquine and @paul121! Hopefully the next PR will be simpler. :-)

**Comment by mstenta** [comment]
Here is the new PR: https://github.com/farmOS/farmOS/pull/854

**Pull request closed without merging**